Merge Shadow Priest Changes

Saeldur opened the pull request:
Fix @seanpeters86 initial work on mind blast charges with voidform
Implement an initial working version of the Dark Thoughts passive including some tracking
Allow mindblast to be instant cast during the dark thoughts passive, but only during mind flay or sear. Thank you @vituscze for the assistance.
Slight APL adjustments to accommodate the passive

Saeldur committed:
Change how voidform removes the charges of mindblast to avoid errors and
attempt to match ingame

Saeldur committed:
Adjust voidform charges logic to avoid assertion fail.
Add basic implementation of dark thoughts.
Cast while channeling is currently not fully functional.

Saeldur committed:
Make dark thoughts proc on sear ticks.

Saeldur committed:
Fix instant cast mindflays not being castable while channeling
Fix formatting to 2spaces per tab

Saeldur committed:
Add proc tracking, APL adjustment, comments
Track procs of DT from sources
Add instant mindblasts to default APL in a basic form
Add TODOs and a few comments

Saeldur committed:
Add Blood Elf Racial to APL
Use arcane torrent prepull for the insanity

scamille commented:
Does dark thoughts have 2% independent proc chance for each dot up, or are they cumulative as in your code?

Otherwise looks good. Would be nice if you could run clang-format on those changes, but not strictly necessary.

The usable while casting expression can be a bit more optimized for performance later, but that's not that important now.

Saeldur commented:
It's currently unknown precisely how dark thoughts works, it would logically be cumulative for each dot up per target but there hasn't been enough verification on that. This is kind of just an initial implementation to see how it works/if it works properly to be easier to change later after more information is known. That's why I threw in a bunch of TODOs around some of the proc.
The usable while casting definitely needs to be optimized so that it can return false early as possible but this was just an initial implementation.
Sorry for the messed up formatting, my vstudio had issues with having clang-format run automatically so I had to disable it (Most of the messed up format is from vstudio trying to automatically format stuff for me).

Saeldur committed:
Run Clang Format To Tidy Files
Also change _charges to charges as the _ is unnecessary

seanpeters86 commented on engine/class_modules/priest/sc_priest_shadow.cpp:
```diff
@@ -316,6 +348,24 @@ struct mind_flay_t final : public priest_spell_t
+
+    if ( rng().roll( priest().specs.dark_thoughts->effectN( 1 ).percent() * dots ) )
+    {
+      // sim->print_log("{} rolled DT Proc success from flay with {} chance with {} dots", *player,
```
can you swap this to just be a debug log? that way we can leave it in uncommented and still see it when debug is on

Saeldur replied:
Added to https://github.com/simulationcraft/simc/pull/5203/commits/cddecfc83431b3ddd487cec96d21750b5d466fed alongside of one for mind sear.

seanpeters86 commented on engine/class_modules/priest/sc_priest_shadow.cpp:
```diff
@@ -1553,6 +1614,38 @@ struct vampiric_embrace_t final : public priest_buff_t<buff_t>
+{
+  dark_thoughts_t( priest_t& p ) : base_t( p, "dark_thoughts", p.find_spell( 341205 ) )
+  {
+    this->set_max_stack( 5 );
```
is this not in spelldata anywhere?

Saeldur replied:
Answered in next conversation

seanpeters86 commented on engine/class_modules/priest/sc_priest_shadow.cpp:
```diff
@@ -1553,6 +1614,38 @@ struct vampiric_embrace_t final : public priest_buff_t<buff_t>
+  dark_thoughts_t( priest_t& p ) : base_t( p, "dark_thoughts", p.find_spell( 341205 ) )
+  {
+    this->set_max_stack( 5 );
+    this->set_duration( timespan_t::from_seconds( 6 ) );
```
same as above, this should either be in spell data or duration should be pulled automagically

Saeldur replied:
Testing without manually setting these does not appear to be getting it from the spelldata, it appears to not currently be in there or I'm missing something.

Updated the p.find_spell to instead get by name however in https://github.com/simulationcraft/simc/pull/5203/commits/cddecfc83431b3ddd487cec96d21750b5d466fed

seanpeters86 commented on engine/class_modules/priest/sc_priest_shadow.cpp:
```diff
@@ -2163,27 +2262,46 @@ void priest_t::generate_apl_shadow()
+                    "Use Shadow Crash on CD unless there are adds incoming." );
+  main->add_action(
+      this, "Mind Sear",
+      "target_if=spell_targets.mind_sear>1&buff.dark_thoughts.up,chain=1,interrupt_immediate=1,interrupt_if=ticks>=2" );
```
Can we add comments to these more complex lines to the APL?

```suggestion
      "target_if=spell_targets.mind_sear>1&buff.dark_thoughts.up,chain=1,interrupt_immediate=1,interrupt_if=ticks>=2", "Use Mind Sear to consume any Dark Thought procs on AoE" );
```

Saeldur replied:
Added with a TODO statement in https://github.com/simulationcraft/simc/pull/5203/commits/cddecfc83431b3ddd487cec96d21750b5d466fed

seanpeters86 commented on engine/class_modules/priest/sc_priest_shadow.cpp:
```diff
@@ -2163,27 +2262,46 @@ void priest_t::generate_apl_shadow()
+      "target_if=spell_targets.mind_sear>1&buff.dark_thoughts.up,chain=1,interrupt_immediate=1,interrupt_if=ticks>=2" );
+  main->add_action( this, "Mind Flay",
+                    "if=buff.dark_thoughts.up&variable.dots_up,chain=1,interrupt_immediate=1,interrupt_if=ticks>=2&"
+                    "cooldown.void_bolt.up" );
```
same as above

Saeldur replied:
Added with a TODO statement in https://github.com/simulationcraft/simc/pull/5203/commits/cddecfc83431b3ddd487cec96d21750b5d466fed

Saeldur committed:
Update with suggestions
Adjust usable during current cast to use spell ids over name and
re-order to save cpu usage.
Add the debug events that were commented out as debug events for procs
from sear and flay.
Add a couple of generic comments.
Add comment to APL to help user understand, and also added TODO for
checking how they actually perform vs other options.
Change Dark Thoughts to get the spell data from the name instead of id.
Change Dark Thoughts to duration extend type instead of Extend, which
should be the correct behaviour.

Saeldur commented:
Updated with all of the suggestions
